File: python/zoned_time_lite/dst_detector.py
"""
Platform-based DST detection using zoneinfo APIs

This module uses Python's timezone database via zoneinfo
to detect DST status, ensuring always-current DST rules.
"""

from __future__ import annotations

from datetime import datetime, timezone
from zoneinfo import ZoneInfo

from . import DEFAULT_TIMEZONE
from .timezone_registry import get_timezone_metadata, validate_platform_timezone
from .validator import validate_date, validate_timezone


def is_dst(date: datetime, tz: str = DEFAULT_TIMEZONE) -> bool:
    """
    Check if a date is in Daylight Saving Time for a given timezone.

    Uses platform timezone databases via zoneinfo for accurate, 
    always-current DST detection. Supports all major timezones with automatic 
    fallback to offset-based detection when abbreviations are unavailable.

    Args:
        date: The date to check (must be a valid datetime object)
        tz: IANA timezone identifier (defaults to 'Europe/London')

    Returns:
        True if the date is in DST for the specified timezone

    Raises:
        ValueError: If date is invalid (NaT) or outside supported range (1970-2100)
        ValueError: If timezone is not supported or unavailable on platform

    Example:
        >>> from datetime import datetime
        >>> summer_date = datetime(2024, 7, 15, 12, 0, 0, tzinfo=timezone.utc)
        >>> winter_date = datetime(2024, 1, 15, 12, 0, 0, tzinfo=timezone.utc)
        >>> 
        >>> is_dst(summer_date, 'Europe/London')  # True (BST)
        >>> is_dst(winter_date, 'Europe/London')  # False (GMT) 
        >>> is_dst(summer_date, 'Asia/Tokyo')     # False (no DST)
    """
    validate_date(date)
    validate_timezone(tz)

    metadata = get_timezone_metadata(tz)
    validate_platform_timezone(tz)

    # If timezone doesn't have DST, it's never in DST
    if metadata.dst_offset is None:
        return False

    # Convert to target timezone
    zone = ZoneInfo(tz)
    localized_date = date.astimezone(zone)

    # Check if DST is active using the dst() method
    dst_info = localized_date.dst()

    if dst_info is None:
        return False

    # dst() returns a timedelta - if it's non-zero, DST is active
    return dst_info.total_seconds() > 0


def is_bst(date: datetime) -> bool:
    """
    Check if a date is in British Summer Time (convenience function for Europe/London).

    This is a convenience wrapper around is_dst specifically for the Europe/London timezone.
    Returns True when London time is in British Summer Time (BST, UTC+1), False when 
    in Greenwich Mean Time (GMT, UTC+0).

    Args:
        date: The date to check (must be a valid datetime object)

    Returns:
        True if the date is in BST (British Summer Time)

    Raises:
        ValueError: If date is invalid (NaT) or outside supported range (1970-2100)

    Example:
        >>> from datetime import datetime, timezone
        >>> summer_date = datetime(2024, 7, 15, 12, 0, 0, tzinfo=timezone.utc)  # July
        >>> winter_date = datetime(2024, 1, 15, 12, 0, 0, tzinfo=timezone.utc)  # January
        >>> 
        >>> is_bst(summer_date)  # True (BST period)
        >>> is_bst(winter_date)  # False (GMT period)
        >>> 
        >>> # Equivalent to:
        >>> is_dst(summer_date, 'Europe/London')  # True
        >>> is_dst(winter_date, 'Europe/London')  # False
    """
    return is_dst(date, "Europe/London")

File: src/dst-detector.ts
/**
 * Platform-based DST detection using Intl APIs
 *
 * This module uses the platform's timezone database via Intl.DateTimeFormat
 * to detect DST status, ensuring always-current DST rules.
 */

import { getTimezoneMetadata } from "./timezone-registry.js";
import {
  validateDate,
  validateTimezone,
  validatePlatformTimezone,
} from "./utils/validation.js";
import { getTimezoneOffset } from "./utils/date-utils.js";
import { timezoneDetector } from "./timezone-detector.js";

/**
 * Check if a date is in Daylight Saving Time for a given timezone
 *
 * Uses platform timezone databases via Intl.DateTimeFormat for accurate,
 * always-current DST detection. Supports all major timezones with automatic
 * fallback to offset-based detection when abbreviations are unavailable.
 *
 * When no timezone is provided, automatically detects the user's timezone.
 *
 * @param date - The date to check (must be a valid Date object)
 * @param timezone - IANA timezone identifier (optional, auto-detects if omitted)
 * @returns `true` if the date is in DST for the specified or detected timezone
 *
 * @throws {Error} If date is invalid (NaN) or outside supported range (1970-2100)
 * @throws {Error} If timezone is not supported or unavailable on platform
 * @throws {TimezoneDetectionError} If timezone detection fails and fallback is invalid
 *
 * @example
 * ```typescript
 * const summerDate = new Date('2024-07-15T12:00:00Z');
 * const winterDate = new Date('2024-01-15T12:00:00Z');
 *
 * // With explicit timezone
 * console.log(isDST(summerDate, 'Europe/London'));  // true (BST)
 * console.log(isDST(winterDate, 'Europe/London'));  // false (GMT)
 * console.log(isDST(summerDate, 'Asia/Tokyo'));     // false (no DST)
 *
 * // With auto-detection (uses user's timezone)
 * console.log(isDST(summerDate));  // true/false based on user's timezone
 * console.log(isDST(winterDate));  // true/false based on user's timezone
 * ```
 */
export function isDST(date: Date, timezone?: string): boolean {
  validateDate(date);

  // Use auto-detection if no timezone provided
  const effectiveTimezone = timezone ?? timezoneDetector.getDetectedTimezone();
  validateTimezone(effectiveTimezone);

  const metadata = getTimezoneMetadata(effectiveTimezone);
  validatePlatformTimezone(effectiveTimezone);

  // If timezone doesn't have DST, it's never in DST
  if (!metadata.dstOffset) {
    return false;
  }

  // Use Intl.DateTimeFormat to get timezone information
  const formatter = new Intl.DateTimeFormat("en", {
    timeZone: effectiveTimezone,
    timeZoneName: "short",
  });

  const parts = formatter.formatToParts(date);
  const timeZoneName = parts.find(
    (part) => part.type === "timeZoneName"
  )?.value;

  if (!timeZoneName) {
    // Fallback: compare offset with expected DST offset
    return isDSTByOffset(date, effectiveTimezone);
  }

  // Check if the timezone name indicates DST
  return isDSTAbbreviation(timeZoneName, effectiveTimezone, date);
}

/**
 * Check if a date is in British Summer Time (convenience function for Europe/London)
 *
 * This is a convenience wrapper around isDST specifically for the Europe/London timezone.
 * Returns true when London time is in British Summer Time (BST, UTC+1), false when
 * in Greenwich Mean Time (GMT, UTC+0).
 *
 * @param date - The date to check (must be a valid Date object)
 * @returns `true` if the date is in BST (British Summer Time)
 *
 * @throws {Error} If date is invalid (NaN) or outside supported range (1970-2100)
 *
 * @example
 * ```typescript
 * const summerDate = new Date('2024-07-15T12:00:00Z');  // July
 * const winterDate = new Date('2024-01-15T12:00:00Z');  // January
 *
 * console.log(isBST(summerDate));  // true (BST period)
 * console.log(isBST(winterDate));  // false (GMT period)
 *
 * // Equivalent to:
 * console.log(isDST(summerDate, 'Europe/London'));  // true
 * console.log(isDST(winterDate, 'Europe/London'));  // false
 * ```
 */
export function isBST(date: Date): boolean {
  return isDST(date, "Europe/London");
}

/**
 * Determine if a timezone abbreviation indicates DST
 * @param abbreviation - The timezone abbreviation from Intl
 * @param timezone - The timezone identifier
 * @returns True if the abbreviation indicates DST
 */
function isDSTAbbreviation(
  abbreviation: string,
  timezone: string,
  date: Date
): boolean {
  const metadata = getTimezoneMetadata(timezone);

  // Check against known DST abbreviations
  if (
    metadata.preferredAbbreviations?.dst &&
    abbreviation === metadata.preferredAbbreviations.dst
  ) {
    return true;
  }

  // Check against known standard abbreviations
  if (
    metadata.preferredAbbreviations?.standard &&
    abbreviation === metadata.preferredAbbreviations.standard
  ) {
    return false;
  }

  // Handle GMT+offset format (common fallback from Intl)
  if (abbreviation.startsWith("GMT")) {
    const offsetMatch = abbreviation.match(/GMT([+-])(\d{1,2})/);
    if (offsetMatch) {
      const sign = offsetMatch[1] === "+" ? 1 : -1;
      const hours = parseInt(offsetMatch[2] ?? "0");
      const offsetMinutes = sign * hours * 60;

      // Compare with expected DST and standard offsets
      if (metadata.dstOffset && offsetMinutes === metadata.dstOffset) {
        return true;
      }
      if (offsetMinutes === metadata.standardOffset) {
        return false;
      }
    }
  }

  // Fallback to offset-based detection for unknown abbreviations
  return isDSTByOffset(date, timezone);
}

/**
 * Fallback DST detection by comparing current offset with expected offsets
 * @param date - The date to check
 * @param timezone - The timezone identifier
 * @returns True if the current offset matches the DST offset
 */
function isDSTByOffset(date: Date, timezone: string): boolean {
  const metadata = getTimezoneMetadata(timezone);

  // If no DST offset defined, never in DST
  if (!metadata.dstOffset) {
    return false;
  }

  // Get the current offset for this date/timezone
  const currentOffset = getTimezoneOffset(date, timezone);

  // Compare with expected DST offset
  return currentOffset === metadata.dstOffset;
}
